File: src/Daisi.Llama/Inference/Sampler.cs
namespace Daisi.Llama.Inference;

/// <summary>
/// Transforms raw logits into a selected token ID through a configurable pipeline:
/// repetition penalty → temperature → top-k → top-p → softmax → sample/argmax.
/// </summary>
public sealed class Sampler
{
    private readonly Random _rng;

    public Sampler(int? seed = null)
    {
        _rng = seed.HasValue ? new Random(seed.Value) : new Random();
    }

    /// <summary>
    /// Sample a token ID from logits using the given parameters.
    /// </summary>
    public int Sample(ReadOnlySpan<float> logits, GenerationParams p, ReadOnlySpan<int> previousTokens)
    {
        int vocabSize = logits.Length;

        // Copy logits so we can mutate
        Span<float> work = vocabSize <= 8192
            ? stackalloc float[vocabSize]
            : new float[vocabSize];
        logits.CopyTo(work);

        // 1. Repetition penalty
        if (p.RepetitionPenalty != 1.0f)
            ApplyRepetitionPenalty(work, previousTokens, p.RepetitionPenalty);

        // 2. Greedy shortcut
        if (p.Temperature == 0)
            return ArgMax(work);

        // 3. Temperature
        float invTemp = 1.0f / p.Temperature;
        for (int i = 0; i < vocabSize; i++)
            work[i] *= invTemp;

        // 4. Top-k
        if (p.TopK > 0 && p.TopK < vocabSize)
            ApplyTopK(work, p.TopK);

        // 5. Softmax
        Softmax(work);

        // 6. Top-p (nucleus) — applied after softmax on probabilities
        if (p.TopP > 0 && p.TopP < 1.0f)
            ApplyTopP(work, p.TopP);

        // 7. Weighted random sample
        return SampleFromDistribution(work);
    }

    private static void ApplyRepetitionPenalty(Span<float> logits, ReadOnlySpan<int> previousTokens, float penalty)
    {
        for (int i = 0; i < previousTokens.Length; i++)
        {
            int tokenId = previousTokens[i];
            if ((uint)tokenId >= (uint)logits.Length) continue;

            if (logits[tokenId] > 0)
                logits[tokenId] /= penalty;
            else
                logits[tokenId] *= penalty;
        }
    }

    private static int ArgMax(ReadOnlySpan<float> values)
    {
        int best = 0;
        float bestVal = values[0];
        for (int i = 1; i < values.Length; i++)
        {
            if (values[i] > bestVal)
            {
                bestVal = values[i];
                best = i;
            }
        }
        return best;
    }

    private static void ApplyTopK(Span<float> logits, int k)
    {
        // Find the k-th largest value using partial sort
        int n = logits.Length;

        // Find top-k threshold
        float kthValue = FindKthLargest(logits, k);

        // Zero out everything below threshold
        for (int i = 0; i < n; i++)
        {
            if (logits[i] < kthValue)
                logits[i] = float.NegativeInfinity;
        }
    }

    private static float FindKthLargest(ReadOnlySpan<float> values, int k)
    {
        // Simple approach: collect top-k values
        Span<float> topK = k <= 1024
            ? stackalloc float[k]
            : new float[k];
        topK.Fill(float.NegativeInfinity);

        for (int i = 0; i < values.Length; i++)
        {
            float v = values[i];
            if (v > topK[k - 1])
            {
                topK[k - 1] = v;
                // Bubble up
                for (int j = k - 2; j >= 0; j--)
                {
                    if (topK[j + 1] > topK[j])
                        (topK[j], topK[j + 1]) = (topK[j + 1], topK[j]);
                    else
                        break;
                }
            }
        }

        return topK[k - 1];
    }

    private static void Softmax(Span<float> values)
    {
        float max = float.NegativeInfinity;
        for (int i = 0; i < values.Length; i++)
            if (values[i] > max) max = values[i];

        float sum = 0;
        for (int i = 0; i < values.Length; i++)
        {
            values[i] = MathF.Exp(values[i] - max);
            sum += values[i];
        }

        float invSum = 1.0f / sum;
        for (int i = 0; i < values.Length; i++)
            values[i] *= invSum;
    }

    private static void ApplyTopP(Span<float> probs, float p)
    {
        // Build (index, prob) pairs and sort descending
        int n = probs.Length;
        Span<(int idx, float prob)> sorted = n <= 8192
            ? stackalloc (int, float)[n]
            : new (int, float)[n];

        for (int i = 0; i < n; i++)
            sorted[i] = (i, probs[i]);

        // Sort descending by probability
        sorted.Sort((a, b) => b.prob.CompareTo(a.prob));

        // Find cumulative threshold
        float cumulative = 0;
        int cutoff = n;
        for (int i = 0; i < n; i++)
        {
            cumulative += sorted[i].prob;
            if (cumulative > p)
            {
                cutoff = i + 1;
                break;
            }
        }

        // Zero out tokens beyond cutoff
        for (int i = cutoff; i < n; i++)
            probs[sorted[i].idx] = 0;

        // Renormalize
        float sum = 0;
        for (int i = 0; i < n; i++)
            sum += probs[i];
        if (sum > 0)
        {
            float invSum = 1.0f / sum;
            for (int i = 0; i < n; i++)
                probs[i] *= invSum;
        }
    }

    private int SampleFromDistribution(ReadOnlySpan<float> probs)
    {
        float r = (float)_rng.NextDouble();
        float cumulative = 0;
        for (int i = 0; i < probs.Length; i++)
        {
            cumulative += probs[i];
            if (r < cumulative)
                return i;
        }
        // Fallback: return last non-zero
        for (int i = probs.Length - 1; i >= 0; i--)
            if (probs[i] > 0) return i;
        return 0;
    }
}

File: src/webgpu/src/model/sampler.ts
/**
 * Token sampler: temperature, top-k, top-p, repetition penalty.
 * Operates on CPU (logits read back from GPU via mapAsync).
 */

export interface SamplerOptions {
  temperature?: number;
  topK?: number;
  topP?: number;
  repetitionPenalty?: number;
  seed?: number;
}

const DEFAULT_OPTIONS: Required<SamplerOptions> = {
  temperature: 0.7,
  topK: 40,
  topP: 0.9,
  repetitionPenalty: 1.1,
  seed: 0,
};

export class Sampler {
  private options: Required<SamplerOptions>;
  private rng: () => number;

  constructor(options?: SamplerOptions) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
    // Simple seeded RNG (xorshift32)
    if (this.options.seed > 0) {
      let state = this.options.seed;
      this.rng = () => {
        state ^= state << 13;
        state ^= state >> 17;
        state ^= state << 5;
        return (state >>> 0) / 0x100000000;
      };
    } else {
      this.rng = Math.random;
    }
  }

  /**
   * Sample a token ID from logits.
   * @param logits - Raw logits array [vocabSize]
   * @param previousTokens - Recent token IDs for repetition penalty
   */
  sample(logits: Float32Array, previousTokens?: number[]): number {
    const { temperature, topK, topP, repetitionPenalty } = this.options;

    // Apply repetition penalty
    if (previousTokens && previousTokens.length > 0 && repetitionPenalty !== 1.0) {
      const seen = new Set(previousTokens.slice(-64)); // Last 64 tokens
      for (const id of seen) {
        if (id < logits.length) {
          if (logits[id] > 0) {
            logits[id] /= repetitionPenalty;
          } else {
            logits[id] *= repetitionPenalty;
          }
        }
      }
    }

    // Greedy (temperature = 0)
    if (temperature === 0 || temperature < 1e-6) {
      return argmax(logits);
    }

    // Apply temperature
    for (let i = 0; i < logits.length; i++) {
      logits[i] /= temperature;
    }

    // Build candidates sorted by logit (descending)
    const indices = new Uint32Array(logits.length);
    for (let i = 0; i < indices.length; i++) indices[i] = i;
    indices.sort((a, b) => logits[b] - logits[a]);

    // Top-K filtering
    let k = Math.min(topK, logits.length);

    // Softmax over top-k candidates
    let maxLogit = logits[indices[0]];
    const probs = new Float32Array(k);
    let sum = 0;
    for (let i = 0; i < k; i++) {
      probs[i] = Math.exp(logits[indices[i]] - maxLogit);
      sum += probs[i];
    }
    for (let i = 0; i < k; i++) probs[i] /= sum;

    // Top-P (nucleus) filtering
    let cumProb = 0;
    let cutoff = k;
    for (let i = 0; i < k; i++) {
      cumProb += probs[i];
      if (cumProb >= topP) {
        cutoff = i + 1;
        break;
      }
    }

    // Renormalize after top-p cutoff
    sum = 0;
    for (let i = 0; i < cutoff; i++) sum += probs[i];
    for (let i = 0; i < cutoff; i++) probs[i] /= sum;

    // Weighted random sampling
    const r = this.rng();
    cumProb = 0;
    for (let i = 0; i < cutoff; i++) {
      cumProb += probs[i];
      if (r <= cumProb) return indices[i];
    }

    return indices[0];
  }
}

function argmax(arr: Float32Array): number {
  let maxIdx = 0;
  let maxVal = arr[0];
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > maxVal) {
      maxVal = arr[i];
      maxIdx = i;
    }
  }
  return maxIdx;
}
